common/src/core/java/xyz/wagyourtail/jsmacros/core/service/ServiceManager.java

```diff
@@ -1,26 +1,61 @@
 package xyz.wagyourtail.jsmacros.core.service;
 
 import com.google.common.collect.ImmutableSet;
+import it.unimi.dsi.fastutil.objects.Object2LongArrayMap;
+import it.unimi.dsi.fastutil.objects.Object2LongMap;
 import xyz.wagyourtail.Pair;
 import xyz.wagyourtail.jsmacros.core.Core;
 import xyz.wagyourtail.jsmacros.core.config.CoreConfigV2;
 import xyz.wagyourtail.jsmacros.core.language.EventContainer;
 
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Set;
+import java.util.Timer;
+import java.util.TimerTask;
 
 /**
  * @author Wagyourtail
  * @since 1.6.3
  */
 public class ServiceManager {
+    protected final Timer timer = new Timer("JSMacros-Service-Filechange-Listener", true);
+    protected boolean reloadOnModify;
+    protected final Object2LongMap<String> lastModifiedMap = new Object2LongArrayMap<>();
+    protected final Set<String> crashedServices = new HashSet<>();
+    
     protected final Core<?, ?> runner;
     protected final Map<String, Pair<ServiceTrigger, EventContainer<?>>> registeredServices = new LinkedHashMap<>();
 
     public ServiceManager(Core<?, ?> runner) {
         this.runner = runner;
+        timer.scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                if (!reloadOnModify) {
+                    return;
+                }
+                for (Map.Entry<String, Pair<ServiceTrigger, EventContainer<?>>> service : registeredServices.entrySet()) {
+                    String file = service.getValue().getT().file;
+                    String name = service.getKey();
+                    // Only restart enabled and running services, i.e. services that are supposed to be running
+                    // If the service is not running because it crashed, try to restart it
+                    if (isEnabled(name) && (isRunning(name) || crashedServices.contains(name))) {
+                        long lastModified = runner.config.macroFolder.getAbsoluteFile().toPath().resolve(file).toFile().lastModified();
+                        if (!lastModifiedMap.containsKey(file)) {
+                            lastModifiedMap.put(file, lastModified);
+                            // Just assume that if the file was changed so was its content. Otherwise, use Adler-32 or MD5 checksum
+                        } else if (lastModifiedMap.getLong(file) != lastModified) {
+                            lastModifiedMap.put(file, lastModified);
+                            crashedServices.remove(name);
+                            restartService(name);
+                        }
+                    }
+                }
+            }
+        }, 0, 500);
```

Comment:
A crashed servcie will be, like any other service, only be restarted when its file changes. The idea is that when you mess up and the service crashes for some reason, which can happend quite frequently when writing a new service, you can just fix the problem, save the file and the service will be restarted with the new, hopefully fixed code. 
I think you meant auto restarting the script every time it crashes?
I can change the timer to use the tick event instead.
ServiceStatus is already a mess because the enum contains 2 independent values, so I don't like the idea of adding any other state.

Reply:
I changed it to use TickBasedEvents, but with the same delay of 500ms (10 ticks). It's also optional, you can toggle it in the settings and it's off by default.

Reply:
> off by default.

ok, fine